### ISSUE TITLE
Would be nice for Firewalld to be able to list rich rules and direct rules

### ISSUE THREAD

**Issue by swelljoe** [webmin/webmin#1622]
fail2ban adds rules using either rich rules or direct rules when interacting with firewalld. Currently the Webmin firewalld module doesn't seem to know these rules exist, so users are left in the dark about what's happening, unless they hit the command line.

This is a thread on the subject from a user that did some digging: https://forum.virtualmin.com/t/my-firewalld-isnt-working-what-is-the-correct-firewallbackend-please-check-yours-for-me/114722/11

Seems like RHEL systems use rich rules, while apparently Debian uses direct rules. I bet we could consolidate them. Ilia has been digging into firewalld and fail2ban, so maybe we can make things act more consistent across distros without too much disruption. I think that'd be a nice thing.

**Comment by iliajie** [comment]
Yeah, that would be an awesome feature to have!

**Comment by iliajie** [comment]
The cool part would be is to combine it into something that `firewall-cmd --direct --get-all-rules` outputs and `ipset list`:

```
ipv4 filter INPUT_direct 0 -p tcp -m multiport --dports 10000 -m set --match-set f2b-webmin-auth src -j REJECT --reject with icmp-port-unreachable
                                                                                 |
                                                                                 | Name: f2b-webmin-auth
                                                                                 | Type: hash:ip
                                                                                 | Revision: 5
                                                                                 | Header: family inet hashsize 1024 maxelem 65536 timeout 0 bucketsize 12 initval 0x206913c7
                                                                                 | Size in memory: 264
                                                                                 | References: 1
                                                                                 | Number of entries: 1
                                                                                 | Members:
                                                                                 | 10.211.55.2 timeout 0

```

**Comment by jcameron** [comment]
Are "rich rules" basically just the same as regular IPtables rules?

**Comment by iliajie** [comment]
Alright, in continuation to new features added earlier to Fail2Ban and FirewallD.

@swelljoe Joe, how does it look like for you for starters?

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/4426533/162746508-84e3f1a0-a345-4a19-9923-ee8b9784dce1.png">

**Comment by MrPeteH** [comment]
Y'all don't mess around :)

I've been reading up on Rich Rules. They too can reference ipSets ...
AFAIK it's simply a more complete UI/spec to describe more complex firewall features.
Direct _is_ much closer to iptables

The following is a guess w/ respect to fail2ban+ipSet+RichRules:
* There is not (yet) an F2B action for RichRules+ipSets
* Therefore, to take advantage of the huge speed advantage of ipSets:
   * continue to use ipSets in F2B
   * define the appropriate FirewallD rules using RichRules rather than Direct

Not yet sure if that requires custom code or ???

**Comment by MrPeteH** [comment]
@iliaross My comments/questions on what you've got:
* I see the final example uses an ipSet. My f2b-postfix-sasl set has 2963 entries. How will that work? ;)
* Is there any intention to (eventually) support ability to do per-rule logging? (I am happy to discover how easy it is with Rich Rules: just an extra token set per-rule! https://firewalld.org/documentation/man-pages/firewalld.richlanguage.html

**Comment by swelljoe** [comment]
@jcameron direct rules are basically iptables rules (though parsed through firewalld, I think, so it's not exactly iptables rules). Rich rules are somewhere in between.

**Comment by iliajie** [comment]
At the moment I am finishing work on control over removing rich/direct rules from that listing.

Also, having displayed so many entries is going to be a mess. I think the best would be is to trim everything after 10 IPs.

**Comment by iliajie** [comment]
> Is there any intention to (eventually) support ability to do per-rule logging? (I am happy to discover how easy it is with Rich Rules: just an extra token set per-rule!

That we will take a closer look later.

**Comment by iliajie** [comment]
Alright, more additions. IPs limit by default set to 15. (in example below it's set to 5 for testing purposes):

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/4426533/162838337-3134bc7d-6189-4906-add9-409df64885ed.png">

Access to a list of rules, which Joe initially requested can be done from the main page for each zone:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/4426533/162838653-e7b04e00-9593-484e-a8f5-69fe0e094772.png">

The rules that are part of

**Comment by iliajie** [comment]
The direct rules, which are part of _match sets_ cannot be deleted in FirewallD Rules page. Although, with new release will make it possible to control Fail2Ban jail statuses and actions as well:

<img width="1154" alt="image" src="https://user-images.githubusercontent.com/4426533/162839275-9323e26c-afdf-4b60-aef8-114057fd0fcc.png">

<img width="1150" alt="image" src="https://user-images.githubusercontent.com/4426533/162839434-43b5054e-2220-41d4-98de-6c7e437ead2d.png">

**Comment by MrPeteH** [comment]
Note that ban *timing* is always a key element:
* Per-jail setting and per-IP active timer.
* In the next release, multiplied delays are coming.

My own gut sense:
* individual timers per-IP are not all that important, compared to
* per-jail timing IS important
* And in the next release, in essence where this IP is on the increasing-delay scale.

In other words, who cares the status of the current countdown... but knowing this jail and/or IP are being delayed for 1 hr vs 4 days vs 4 weeks is a pretty big deal. ;)

**Comment by swelljoe** [comment]
I would like to discourage anything in the UI that makes it easy/tempting to micromanage blocking. It's another sort of security theater that can distract from the real work of security. If fail2ban is working correctly and has a reasonable set of rules, nobody should ever be looking at IPs and thinking about what to block and for how long.

UI can guide users into wasting their time or it can guide them into making effective decisions. We shouldn't be encouraging users to fight with individual IPs on the internet. There will always be another IP. The best thing to do is always make the tools effective at spotting trouble and shutting it down automatically without a human having to think about it.

**Comment by jcameron** [comment]
I would agree with Joe here - blocking specific IPs isn't really useful because the attackers will just move. Fail2ban solves this by detecting and blocking IPs dynamically better than a human could, but making it's blocks permanent isn't likely to help much in the long term.

**Comment by iliajie** [comment]
Alright, I will remove it from Fail2Ban and will add a rich rule editor on FirewallD side.

**Comment by MrPeteH** [comment]
YES!
My gut UI sense follows my long-term rule: _hide the complexity._ 
If anything, have a button leading to a (popup?) that allows seeing/scrolling through the current set of specific IP's... yet most won't care.
And we already presume that those who need niggly details can go to a shell prompt and use the appropriate command. :)

**Comment by MrPeteH** [comment]
What WOULD help is perhaps an audit (??) button to check block periods (difference between block and unblock timestamps.) 
_Is the actual period ever less than the configured block period?_ 

I've seen some config errors cause this failure, and it is not easy to detect.

**Comment by iliajie** [comment]
> What WOULD help is perhaps an audit (??) button to check block periods (difference between block and unblock timestamps.)

I think this is all various types of micromanagement that Joe advised against. All of this is possible to do but not at the moment this doesn't seem to be that important.

> If anything, have a button leading to a (popup?) that allows seeing/scrolling through the current set of specific IP's

I have thought of doing it too. I will add this to my todo for the future.

**Comment by MrPeteH** [comment]
re Micromanagement: my priority there is not the details... but simply verification of overall proper operation.

(Honestly, a popup of IP list feels like micromanagement to me ;) ... /bin/sh is just fine for that ;) )

The fail2ban ethos in general:
* For certain classes of log errors
* Ban bad IP's for <X amount of time>
* (And new feature: X can increase)

And the question I raised above about an "audit" is simply a verification...if I compare <unban>-<ban> time, is it significantly different from configured... particularly if less (ie bad guys getting through)

HMM... the following is likely a separate suggestion
I have a hunch we could script a nice little fail2ban performance/analysis table...

Category   Total IP's (since start)    Active Bans    % Not Banned   Cfg Ban Time   Avg Actual Ban Time

where
- Category: filter.d name ;)
- Total IP's in that category
- Active: # currently banned
- % not banned (some IP's don't trigger a ban... lets admin know ... maybe need to reduce limit
- Cfg Ban Time: configured baseline ban time. Multiplier will make this grow of course
- Avg actual ban: (I've done this: select all Ban/Unban. Sort on: IP, timestamp, then subtract unban from prior ban)

hugely helpful...
- are my filters catching anything
- which ones most active
- are attackers using drip techniques to leak in (% not banned...of course could be user errors if tiny)
- is something causing ban times to be smaller than configured...

**Comment by MrPeteH** [comment]
i may have to write that script. after i get my taxes done. back to the Real World. :)

**Comment by Chaots** [comment]
Hey All,

first of all, i am happy with webmin/virtualmin as it is, so changes for me are not a priority desire, i do however want to share my reasons for being able to see / have more control over certain parts of the rules active in FirewallD. I use webmin/virtualmin for machines that act as a routers for friends and my own networks at home. These machines tend to run 1 interface for internet, 3 to 4 internal interfaces(be it physical or vlan's) often a split of home, office, iot and guests, in some cases even more. and finally a vpn tunnel that caries in a few more networks/interfaces. web/virtualmin provides dns, dhcp, and in some cases some web services like nextcloud or a simple run it at home project for the plethora of iot/diy devices we tend to make and play with.

As it is the per interface firewalls work more then well enough, although for some things we need to fall back to cli to make those additional changes we like to have. the FirewallD module is nice enough to not mess this up when adding or changing these zones so there is not a real problem / things are not broken. some examples where we need to fall back to cli:
1) Add/Remove "sources" to a zone. 
-- Example:
# firewall-cmd --add-source=1.1.1.1 --zone=home
# firewall-cmd --add-source=1.1.1.1/28 --zone=iot 
-- Why:
We use this is used in the networks where a /24 subnet is divided and we would like to only expose a port/service to a part of that subnet, for example, drop/allow forward or masquerade.

2} open a port / service only to a single host via rich rules while maintaining a standard set in the zone.
-- example
# firewall-cmd --permanent --zone=External --add-rich-rule='rule family=ipv4 source address=192.168.1.100 port port=25565 protocol=tcp accept'
-- Why?
Sometimes a game server ( minecraft in this case ) is exposed to the public side for friends to be able to join us in our private server. in the example i use minecraft, this could also be the listening port for a vpn deamon, or a custom port for an IoT api/test web page. while this is not the main way to manage access control's, it is certainly an additional layer of prevention.

3) IPv6, dropping or filtering specific services on the the edge with the router. while i currently have no examples for ipv6 in active use, we are gradually getting IPv6 on all sites and it will become a thing. services like SMB, Zabbix-agent, or poor built / old drivers like a marvel raid utility happy expose their ports on IPv6 not always taking into account the scope ( local fe80 vs public 2001:.. )
-- Why
Once again not really security, as systems should have their own local firewall defend the lines and ACL's should make it more fine grained, it does not hurt to have those extra bits of defense in the line.

All in all non of the above is a problem, manually adding these things on the cli works fine. The FirewallD module seems to respect these manual additions well and there are no visible problems after doing the manual additions. It would however be nice to be able to see them listed, even if only as information that can not be changed from the GUI.

At the same time i agree with  @swelljoe and @jcameron and their "statements" about micromanaging. Even though that was in the context of "blocking ip's, these features as option for FirewallD may bring a horde of people trying to fine-tune every bit of access and bring in the requests for every single option that rich/direct rules offer.

The "old" iptables module was great for this though it's multi level / multi dimension way of needing to work was quite a challenge for many people. I would hate it if the current FirewallD module in Webmin would become the same behemoth as the iptables module was/is. Where possible i would almost want these additional options tucked away in a separate module(add-on) something like "FirewallD Advanced". so the basic use on most systems is satisfied as is now in the FirewallD module, and when really needed for those specific needs, one can jump in the advanced module and add their special needs on top of the main FirewallD module.

**Comment by iliajie** [comment]
Thanks for your suggestions!

> IPv6, dropping

Since our last discussion, the ability to block IPs and/or subnets using rich rules has been added. Let’s call it "Quick Block".

> open a port / service only to a single host

This seems like a useful and easy-to-implement feature to quickly allow an IP or service. Let’s call it  "Quick Allow"

> Add/Remove "sources" to a zone.
-- Example:
firewall-cmd --add-source=1.1.1.1 --zone=home
firewall-cmd --add-source=1.1.1.1/28 --zone=iot
-- Why:
We use this is used in the networks where a /24 subnet is divided and we would like to only expose a port/service to a part of that subnet, for example, drop/allow forward or masquerade.

Which category would this fit under? While "Quick Allow" and "Quick Block" are fairly common, I’m not sure what this would be... Maybe we could call it a "Rich Rule Constructor" and build something around that?

**Comment by Chaots** [comment]
the sources section is kind of a unique zone of it's own. Though that zone does only apply to the specific mentioned source addresses it is bound to. it basically excludes the source from the normal zone it would be in. say i would have eth1, with a zone of internal, there i would allow only dhcp, and dns as a default.

I could then make a zone "internal-trusted", link/add a mix of ip's and/or subnets to that zone, and then have a complete different set of rules for the ip's linked. the ip's linked to this zone do NOT inherit / get the rules from the default internal zone as they would based on their interface, every rule, be it allow or deny is unique to the zone the source is linked into. so in essence it's a custom firewall ruleset for a all source's that match.

A usage example for this:
I have 7 of these routers in a group. On the "external" zone linked to the public interface i only provide mail and http(s) services for the world, so those are the only ports listed in this zone.

Though the 7 routers are trusted and all have additional access needs. Now i could create a rich rule for every every additional service, and then do that 7 times for all those routers. If i would have 7 services, that would mean i need to add 49 rich rules. for the external zone.

Instead i have made a zone named "externalrouters". added their ip addresses as source to this zone, i then add the 7 services to this zone, and i call it a day, all seven "sources" now can pass the firewall with their specific zones rules.

you just need to be very aware that using the source marking on an ip, will pull that ip out of any other zone they would normally pass through. as such all rules in the other zone do not apply to this source anymore.

source(s) zone(s)?

**Comment by iliajie** [comment]
Can you share some commands you find useful for your workflow that you currently run manually but could potentially be integrated into the FirewallD module UI?

**Comment by swelljoe** [comment]
Since the original issue has kinda been resolved (viewing rich rules is available in the module), and no follow up in a couple years on anything else brought up, I'll close this. Focused issues are better for getting things done. If something is still needed from this one, open a new one about that specific thing.
